# Assignment 2

## 1

output `small_model.txt`:
```
11 19 
0 i 3 
1 live 1
2 in 1
3 boston 1
4 . 3
5 like 3
6 ants 2
7 honey 2
8 therefore 1
9 too 1
10 ! 1
0 1 -1.098612288668110
0 5 -0.405465108108164
1 2 0.000000000000000
2 3 0.000000000000000
3 4 0.000000000000000
4 0 -1.098612288668110
4 6 -1.098612288668110
4 8 -1.098612288668110
5 6 -1.098612288668110
5 7 -0.405465108108164
6 4 -0.693147180559945
6 5 -0.693147180559945
7 4 -0.693147180559945
7 9 -0.693147180559945
8 0 0.000000000000000
9 10 0.000000000000000
-1
```

bigram probability should be written as:

$ P(w_i | w_{i-1}) = \dfrac{c(w_{i-1}, w_{i})}{c(w_{i-1})}$

# 2

Running `Generator.py`

`python .\Generator.py -f .\kafka_model.txt -s i -n 7`

Output:

`i ca n't eat . now and ,`

Or,

`python .\Generator.py -f kafka_model.txt -s dog -n 5`

Output:

`dog copy out again ! grete`

# 3

Linear Interpolation is

$ P(w_{i-1} w_{i}) =\lambda_{1} P(w_i | w_{i-1}) + \lambda_{2} P(w_i) + \lambda_{3}$

Approximate the entropy as:

$ - \dfrac{1}{N} \log_2 P(w_{i-1}w_{i})$

## (a)

running `run_tester_small_kafka.bat`

I got the output:

`Read 24944 words. Estimated entropy: 13.46`

--

the file `entropy_small_kafka.txt`

`Read 24944 words. Estimated entropy: 13.46`

running `run_tester_kafka_small.bat`



I got the output:

`Read 19 words. Estimated entropy: 10.29`

--

the file `entropy_kafka_small.txt`

`Read 19 words. Estimated entropy: 10.29`

## (b)

*`run_tester_austen_austen.bat`*

I got the output:

`Read 10738 words. Estimated entropy: 6.97`

*`-f austin_model -t guardian_test`*

I got the output:

`Read 871878 words. Estimated entropy: 9.75`

*`run_tester_guardian_guardian.bat`*

I got the output:

`Read 871878 words. Estimated entropy: 6.62`

*`-f guarfian_model -t austen_test`*

I got the output:

`Read 10738 words. Estimated entropy: 6.56`

==

To evaluate these two models, austen and guardian, I ran *`run_tester_austen_kafka.bat`* and *`-f guardian_model -t kafka`*.

I got the output for former one as: `Read 24944 words. Estimated entropy: 7.36`

the latter one as: `Read 24944 words. Estimated entropy: 6.04`

As these results (including the result of austen_test, guardian_test), we can draw a conclusion that **guardian_model is better than austen_model** because of *"The lower the entropy of the test corpus, the better the language model learned from the training corpus."*

# 4

## (a) (b) (e) Batch Gradient

**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |84710|15288| 
| 1 |  0  |  0  |

I looped for 1000 times

<img src="./fig/gd2.png" width="500">

In [1]:
total = 84710 + 15288

accuracy = 84710 / total
pre_0 = 84710 / (84710 + 15288)

recall_0 = 84710 / (84710 + 0)
recall_1 = 0 / (15288 + 0)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.8471169423388468
precision_0:  0.8471169423388468
recall_0:  1.0
recall_1:  0.0


**Accuracy, Precision, Recall**

|   |    |
|---|-----|
| Accuracy |84.7 %|
| Precision_0 |84.7 %|
| Precision_1 |  0  |
| Recall_0 |100 %|
| Recall_1 | 0 % |

## (c) (e) SDG

**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |83831|3242 | 
| 1 | 879 |12046|

10000 loops per epoch, and looped for 10 epochs

<img src="./fig/sto_gd1.png" width="500">

In [2]:
total = 84710 + 15288

accuracy = (83831 + 12046) / total
pre_0 = 83831 / (83831 + 3242)
pre_1 = 12046 / (12046 + 879)

recall_0 = 83831 / (83831 + 879)
recall_1 = 12046 / (12046 + 3242)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("precision_1: ", pre_1)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.9587891757835156
precision_0:  0.9627668737725816
precision_1:  0.9319922630560928
recall_0:  0.9896234210836973
recall_1:  0.7879382522239665


**Accuracy, Precision, Recall**

|   |    |
|---|-----|
| Accuracy |95.9 %|
| Precision_0 |96.3 %|
| Precision_1 | 93.2 %|
| Recall_0 |98.7 %|
| Recall_1 | 78.8 % |

## (d) (e) minibatch
 
**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |80674|2015 | 
| 1 | 4036 |13273|

looped for 10 epochs

<img src="./fig/minibatch.png" width="500">

In [4]:
total = 84710 + 15288

accuracy = (80674 + 13273) / total
pre_0 = 80674 / (80674 + 2015)
pre_1 = 13273 / (13273 + 4036)

recall_0 = 80674 / (80674 + 4036)
recall_1 = 13273 / (13273 + 2015)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("precision_1: ", pre_1)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.9394887897757955
precision_0:  0.975631583402871
precision_1:  0.7668265064417356
recall_0:  0.9523550938496045
recall_1:  0.8681972789115646


**Accuracy, Precision, Recall**

|   |    |
|---|-----|
| Accuracy |93.9 %|
| Precision_0 |97.6 %|
| Precision_1 | 76.7 %|
| Recall_0 |95.2 %|
| Recall_1 | 86.8 % |

## (f)

I add two features.

1. If the word includes the uppercase letter. (Includes uppercase: 1)
2. If the word includes the digit. (Includes digit: 1)

### Using Batch Gradient

**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |81532|14014 | 
| 1 |3178 |1274|


<img src="./fig/gd_f1.png" width="500">

In [5]:
total = 84710 + 15288

accuracy = (81532 + 1274) / total
pre_0 = 81534 / (81534 + 14014)
pre_1 = 1274 / (3178 + 1274)

recall_0 = 81534 / (81534 + 3178)
recall_1 = 1274 / (1274 + 14014)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("precision_1: ", pre_1)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.8280765615312307
precision_0:  0.8533302633231464
precision_1:  0.2861635220125786
recall_0:  0.9624846538861083
recall_1:  0.08333333333333333


**Accuracy, Precision, Recall**

|   |Before|After|
|---|-----|-----|
| Accuracy |84.7 %|82.8 %|
| Precision_0 |84.7 %|85.3 %|
| Precision_1 |  0  | 28.6 %|
| Recall_0 |100 %| 96.2 %|
| Recall_1 | 0 % | 8.3 %|

### Using SGD


**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |83604|2764 | 
| 1 |1106 |12524|


<img src="./fig/sto_gd_f2.png" width="500">

In [6]:
total = 84710 + 15288

accuracy = (83604 + 12524) / total
pre_0 = 83604 / (83604 + 2764)
pre_1 = 12524 / (1106 + 12524)

recall_0 = 83604 / (83604 + 1106)
recall_1 = 12524 / (12524 + 2764)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("precision_1: ", pre_1)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.9612992259845197
precision_0:  0.9679974064468322
precision_1:  0.9188554658840793
recall_0:  0.9869436902372801
recall_1:  0.8192046049188907


**Accuracy, Precision, Recall**

|   |Before|After|
|---|-----|-----|
| Accuracy |95.9 %|96.1 %|
| Precision_0 |96.3 %|96.8 %|
| Precision_1 | 93.2 %|91.9 %|
| Recall_0 |98.7 %|98.7 %|
| Recall_1 | 78.8 % |81.9 %|

### Using minibatch

**Real vs Predicted**

|   |  0  |  1  |
|---|-----|-----|
| 0 |80673|2011 | 
| 1 | 4037 |13277|

<img src="./fig/minibatch_f1.png" width="500">

In [8]:
total = 84710 + 15288

accuracy = (80673 + 13277) / total
pre_0 = 80673 / (80673 + 2011)
pre_1 = 13277 / (13277 + 4037)

recall_0 = 80673 / (80673 + 4037)
recall_1 = 13277 / (13277 + 2011)

print("accuracy: ", accuracy)
print("precision_0: ", pre_0)
print("precision_1: ", pre_1)
print("recall_0: ", recall_0)
print("recall_1: ", recall_1)

accuracy:  0.9395187903758075
precision_0:  0.9756784867689033
precision_1:  0.7668360864040661
recall_0:  0.9523432888679022
recall_1:  0.8684589220303506


**Accuracy, Precision, Recall**

|   |Before|After|
|---|-----|-----|
| Accuracy |93.9 %|94.0 %|
| Precision_0 |97.6 %|97.7 %|
| Precision_1 | 76.7 %|76.7 %|
| Recall_0 |95.2 %|95.2 %|
| Recall_1 | 86.8 % | 86.8 %|